mongosql/src/algebrizer/test.rs

```diff
@@ -7454,20 +7454,256 @@ mod limit_or_offset_clause {
 }
 
 mod set_query {
-    use super::{catalog, mir_source_bar, mir_source_foo, AST_QUERY_BAR, AST_QUERY_FOO};
-    use crate::{ast, mir, mir::schema::SchemaCache, usererror::UserError};
+    use super::{
+        catalog, mir_source_bar, mir_source_foo, AST_QUERY_BAR, AST_QUERY_FOO, AST_SOURCE_BAR,
+        AST_SOURCE_FOO,
+    };
+    use crate::schema::{Document, Schema};
+    use crate::{
+        ast, map, mir, mir::schema::SchemaCache, multimap, schema, set,
+        unchecked_unique_linked_hash_map,
+    };
+    use mongosql_datastructures::binding_tuple::DatasourceName;
 
     test_algebrize!(
-        union_distinct_not_allowed,
+        union_distinct_star,
         method = algebrize_set_query,
-        expected = Err(Error::DistinctUnion),
-        expected_error_code = 3006,
+        expected = Ok(mir::Stage::Project(mir::Project {
+            source: Box::new(mir::Stage::Group(mir::Group {
+                source: Box::new(mir::Stage::Set(mir::Set {
+                    operation: mir::SetOperation::UnionAll,
+                    left: Box::new(mir::Stage::Project(mir::Project {
+                        source: Box::new(mir::Stage::Collection(mir::Collection {
+                            db: "test".into(),
+                            collection: "foo".into(),
+                            cache: SchemaCache::new()
+                        })),
+                        expression: map! {
+                            (DatasourceName::Named("foo".into()), 0u16).into() => mir::Expression::Reference((DatasourceName::Named("foo".into()), 0u16).into())
+                        },
+                        is_add_fields: false,
+                        cache: SchemaCache::new()
+                    })),
+                    right: Box::new(mir::Stage::Project(mir::Project {
+                        source: Box::new(mir::Stage::Collection(mir::Collection {
+                            db: "test".into(),
+                            collection: "bar".into(),
+                            cache: SchemaCache::new()
+                        })),
+                        expression: map! {
+                            (DatasourceName::Named("bar".into()), 0u16).into() => mir::Expression::Reference((DatasourceName::Named("bar".into()), 0u16).into())
+                        },
+                        is_add_fields: false,
+                        cache: SchemaCache::new()
+                    })),
+                    cache: SchemaCache::new()
+                })),
+                keys: vec![
+                    mir::OptionallyAliasedExpr::Aliased(mir::AliasedExpr {
+                        alias: "__groupKey0".to_string(),
+                        expr: mir::Expression::Reference(
+                            (DatasourceName::Named("bar".to_string()), 0u16).into()
+                        )
```

Comment:
```suggestion
                        expr: mir::Expression::Reference(("bar", 0u16).into())
```
We have a `impl<T: Into<Key>> From<T> for ReferenceExpr` which utilizes `impl<D, S> From<(D, S)> for Key where D: Into<DatasourceName>, S: Into<u16>` which utilizes `impl<S> From<S> for DatasourceName where S: Into<String>`.

All of that to say, we ultimately allow converting `(&str, u16)` tuples into `ReferenceExpr`s directly! So you could remove the `DatasourceName::Named("...".to_string())` wherever you create an `Expression::Reference` to help cut down on the size of these tests!

Reply:
Sounds good.  I made this change and added a helper function to hopefully make things easier to follow.